avocado/formatters.py

```diff
@@ -95,139 +105,238 @@ def to_FORMAT(value, field=field, concept=self.concept,
-    an OrderedDict, or a sequence of key-value pairs.
+    The output of a `process_multiple` method must be a tuple of values.
     """
-    def __init__(self, concept=None, keys=None):
+    def __init__(self, concept=None, keys=None, formats=None):
         "Passing in a concept takes precedence over `keys`."
         if not keys and not concept:
             raise ValueError('A concept or sequence of keys are required.')
 
         self.concept = concept
-        self.fields = None
 
         if concept:
-            fields = list(concept.fields.order_by('concept_fields__order'))
-            self.fields = OrderedDict(_unique_keys(fields))
-            self.keys = self.fields.keys()
+            self.concept_fields = tuple(concept.concept_fields
+                                        .select_related('field')
+                                        .order_by('order', 'name'))
+
+            self.fields = _unique_dict(cf.field
+                                       for cf in self.concept_fields)
+            self.field_names = self.fields.keys()
         else:
-            self.keys = keys
+            self.concept_fields = ()
+            self.fields = {}
+            self.field_names = keys
+
+        # Clean formats and split into single vs. multi.
+        self.single_formats = []
+        self.multi_formats = []
+
+        if formats:
+            for fmt in formats:
+                method = getattr(self, 'to_{0}'.format(fmt), None)
+
+                if method:
+                    if getattr(method, 'process_multiple', False):
+                        self.multi_formats.append(method)
+                    else:
+                        self.single_formats.append(method)
 
         # Keep track of fields/concepts that cause an error to prevent
         # logging the exception twice
         self._errors = set()
 
-    def __call__(self, values, preferred_formats=None, **context):
-        if not preferred_formats:
-            preferred_formats = []
+        # Create a bare-bones record class.
+        self._recordclass = namedtuple('record', self.field_names)
 
-        # Create a OrderedDict of the values relative to the
-        # concept fields objects the values represent. This
-        # enables key-based access to the values rather than
-        # relying on position.
-        if not isinstance(values, OrderedDict):
-            if not isinstance(values, (list, tuple)):
-                values = [values]
+    def __call__(self, values, kwargs=None):
+        return self._process(values, kwargs=kwargs)
 
-            values = OrderedDict(zip(self.keys, values))
+    def __contains__(self, choice):
+        return hasattr(self, 'to_{0}'.format(choice))
 
-        # Create list of formats to attempt for multi-value processing.
-        multi_formats = list(preferred_formats)
+    def __unicode__(self):
+        return u'{0}'.format(self.name or self.__class__.__name__)
+
+    def _process(self, values, kwargs=None):
```

Comment:
Can you add a docstring to this method? It took me a while to sort out(at least I think I have) what exactly is going on here and with the interactions with `_process_single` and `_process_multiple` especially with the `skip` logic.